src/main/scala/swiss/dasch/api/MaintenanceEndpointRoutes.scala

```diff
@@ -0,0 +1,137 @@
+ */
+
+package swiss.dasch.api
+
+import swiss.dasch.api.MaintenanceEndpoint.*
+import swiss.dasch.domain.*
+import zio.json.JsonEncoder
+import zio.nio.file
+import zio.nio.file.Files
+import zio.{ Chunk, IO, ZIO }
+
+import java.io.IOException
+
+object MaintenanceEndpointRoutes {
+
+  private def getProjectPath(shortcodeStr: String): ZIO[ProjectService, ApiProblem, file.Path] =
+    ApiStringConverters
+      .fromPathVarToProjectShortcode(shortcodeStr)
+      .flatMap(code =>
+        ProjectService.findProject(code).some.mapError {
+          case Some(e) => ApiProblem.internalError(e)
+          case _       => ApiProblem.projectNotFound(code)
+        }
+      )
+
+  private def saveReport[A](
+      tmpDir: file.Path,
+      name: String,
+      report: A,
+    )(implicit encoder: JsonEncoder[A]
+    ): ZIO[StorageService, Throwable, Unit] =
+    Files.createDirectories(tmpDir / "reports") *>
+      Files.deleteIfExists(tmpDir / "reports" / s"$name.json") *>
+      Files.createFile(tmpDir / "reports" / s"$name.json") *>
+      StorageService.saveJsonFile(tmpDir / "reports" / s"$name.json", report)
+
+  private val needsOriginalsRoute = needsOriginalsEndpoint.implement(imagesOnlyMaybe =>
+    {
+      val imagesOnly = imagesOnlyMaybe.getOrElse(true)
+      val reportName = if (imagesOnly) "needsOriginals_images_only" else "needsOriginals"
+      for {
+        _                 <- ZIO.logInfo(s"Checking for originals")
+        assetDir          <- StorageService.getAssetDirectory()
+        tmpDir            <- StorageService.getTempDirectory()
+        projectShortcodes <- ProjectService.listAllProjects()
+        _                 <- ZIO
+                               .foreach(projectShortcodes)(shortcode =>
+                                 Files
+                                   .walk(assetDir / shortcode.toString)
+                                   .mapZIOPar(8)(originalNotPresent(imagesOnly))
+                                   .filter(identity)
+                                   .as(shortcode)
+                                   .runHead
+                               )
+                               .map(_.flatten)
+                               .flatMap(saveReport(tmpDir, reportName, _))
+                               .zipLeft(ZIO.logInfo(s"Created $reportName.json"))
+                               .logError
+                               .forkDaemon
+      } yield "work in progress"
+    }.logError.mapError(ApiProblem.internalError)
+  )
+
+  private def originalNotPresent(imagesOnly: Boolean)(path: file.Path): IO[IOException, Boolean] = {
+    val assetId = AssetId.makeFromPath(path).map(_.toString).getOrElse("unknown-asset-id")
+    (ZIO.succeed(imagesOnly).negate || FileFilters.isImage(path)) &&
+    Files
+      .list(path.parent.orNull)
+      .map(_.filename.toString)
+      .filter(name => name.endsWith(".orig") && name.startsWith(assetId))
+      .runHead
+      .map(_.isEmpty)
+  }
```

Comment:
to me, the code would be more self explanatory if this was split into two different functions, instead of having the boolean flag as the first parameter. I really had to think hard what `(ZIO.succeed(imagesOnly).negate || FileFilters.isImage(path)) && ...` means, and this cognitive load could maybe be delegated to the function name, right?

Reply:
fixed this and a small bug in 145a479c6587226aec24f2d856260a5018c9b1e3